### PR TITLE
check private property access outside the class #4464

### PR DESCRIPTION
**Pull request by 0xe**
**Description:**

Private fields should be checked for private access.

**BREAKING CHANGE:**

**Related issue (if exists):**

- Closes https://github.com/swc-project/swc/issues/4464

### CODE REVIEW

File: crates/swc_ecma_transforms_compat/src/es2022/private_in_object.rs [hunk 1]
```diff
@@ -367,6 +368,14 @@ impl VisitMut for PrivateInObject {
                 });
             }
 
+            Expr::Member(MemberExpr { span, obj, prop }) => {
+                if let MemberProp::PrivateName(PrivateName { .. }) = *prop {
+                    let error = "Illegal access to private class property.".to_string();
```
Comment (kdy1) [inline crates/swc_ecma_transforms_compat/src/es2022/private_in_object.rs:373]: We don't need `String`, which allocates, here

**Review by kdy1 — CHANGES_REQUESTED** [review]

File: crates/swc_ecma_transforms_compat/src/es2022/private_in_object.rs [hunk 2]
```diff
@@ -367,6 +368,15 @@ impl VisitMut for PrivateInObject {
...
+            Expr::Member(MemberExpr {
+                span,
+                obj: _,
+                prop: MemberProp::PrivateName(PrivateName { .. }),
+            }) => {
+                let error = "Illegal access to private class property.";
+                HANDLER.with(|h| h.struct_span_err(*span, error).emit());
+            }
+
             _ => {}
```
Comment (kdy1) [inline crates/swc_ecma_transforms_compat/src/es2022/private_in_object.rs:380]: Can you add some tests?

Comment (kdy1) [reply crates/swc_ecma_transforms_compat/src/es2022/private_in_object.rs:380]: You can add some `input.js` within proper directory in https://github.com/swc-project/swc/tree/5122a7dc6270cfa3fb44dbd32cf8e33b808eed0e/crates/swc/tests/errors

Comment (kdy1) [reply crates/swc_ecma_transforms_compat/src/es2022/private_in_object.rs:380]: And I think we need more integration tests (in `swc` crate) for valid cases, too.

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thank you!

File: crates/swc_ecma_transforms_compat/src/es2022/private_in_object.rs (hunk 2 above)
Comment (0xe) [reply crates/swc_ecma_transforms_compat/src/es2022/private_in_object.rs:380]: Will add tests.

File: crates/swc/tests/errors/lints/check-private-property/2/private_props.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+class Foo {
```
Comment (kdy1) [inline crates/swc/tests/errors/lints/check-private-property/2/private_props.js:1]: I think this should be in 

https://github.com/swc-project/swc/tree/7d40208377133b0fcd153c5bfcf47f6b489113c1/crates/swc/tests/exec

**Review by kdy1 — CHANGES_REQUESTED** [review]

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - swc_ecma_transforms_compat

**Review by kdy1 — CHANGES_REQUESTED** [review]
Oh.. CI failed

**Review by kdy1 — APPROVED** [review]

**Review by kdy1 — APPROVED** [review]

**Comment by kdy1** [comment]
I'll revert this and publish a new version

**Pull request merged**